pkg/modprovider/module.go

```diff
@@ -0,0 +1,482 @@
+		ctx,
+		urn,
+		moduleInputs,
+		oldOutputs,
+		inferredModule,
+		moduleSource,
+		moduleVersion,
+		providersConfig,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("Failed preparing tofu sandbox: %w", err)
+	}
+
+	logger := newResourceLogger(h.hc, resource.URN(req.GetUrn()))
+
+	destroyErr := tf.Destroy(ctx, logger)
+	if destroyErr != nil {
+		logger.Log(ctx, tfsandbox.Debug, fmt.Sprintf("error running tofu destroy in delete: %v", destroyErr))
+	}
+
+	// Send back empty pb if no error.
+	return &emptypb.Empty{}, destroyErr
+}
+
+func (h *moduleHandler) Read(
+	ctx context.Context,
+	req *pulumirpc.ReadRequest,
+	moduleSource TFModuleSource,
+	moduleVersion TFModuleVersion,
+	inferredModule *InferredModuleSchema,
+	providersConfig map[string]resource.PropertyMap,
+) (*pulumirpc.ReadResponse, error) {
+	if req.Inputs == nil {
+		return nil, fmt.Errorf("Read() is currently only supported for pulumi refresh")
+	}
+
+	urn := urn.URN(req.GetUrn())
+
+	moduleInputs, err := plugin.UnmarshalProperties(req.Inputs, h.marshalOpts())
+	if err != nil {
+		return nil, err
+	}
+
+	oldOutputs, err := plugin.UnmarshalProperties(req.Properties, h.marshalOpts())
+	if err != nil {
+		return nil, err
+	}
+
+	tf, err := h.prepSandbox(
+		ctx,
+		urn,
+		moduleInputs,
+		oldOutputs,
+		inferredModule,
+		moduleSource,
+		moduleVersion,
+		providersConfig,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("Failed preparing tofu sandbox: %w", err)
+	}
+
+	logger := newResourceLogger(h.hc, resource.URN(req.GetUrn()))
+	state, err := tf.Refresh(ctx, logger)
+	if err != nil {
+		return nil, fmt.Errorf("Module refresh failed: %w", err)
+	}
+
+	outputs, err := h.outputs(ctx, tf, state)
+	if err != nil {
+		return nil, err
+	}
+
+	properties, err := plugin.MarshalProperties(outputs, h.marshalOpts())
+	if err != nil {
+		return nil, err
+	}
+
+	return &pulumirpc.ReadResponse{
+		Id:         moduleResourceID,
+		Properties: properties,
+		Inputs:     req.GetInputs(), // inputs never change on refresh
+	}, nil
+}
+
+func (h *moduleHandler) getState(props resource.PropertyMap) (rawState []byte, rawLockFile []byte) {
+	state, ok := props[moduleResourceStatePropName]
+	if !ok {
+		return // empty
+	}
+	stateString := state.StringValue()
+	rawState = []byte(stateString)
+	if lock, ok := props[moduleResourceLockPropName]; ok {
+		if lock.IsSecret() {
+			lock = lock.SecretValue().Element
+		}
+		lockString := lock.StringValue()
+		rawLockFile = []byte(lockString)
+	}
+	return
+}
+
+func (*moduleHandler) marshalOpts() plugin.MarshalOptions {
+	return plugin.MarshalOptions{
+		KeepUnknowns:  true,
+		KeepSecrets:   true,
+		KeepResources: true,
+
+		// If there are any resource.NewOutputProperty values in old inputs with dependencies, this setting
+		// will ignore the dependencies and remove these values in favor of simpler Computed or Secret values.
+		//
+		// TODO need to figure out if this is actually sufficient, or else the handler needs to extract these
```

Comment:
Would you know off the bat @justinvp or I can run a quick test. We have a test for dependency tracking in the codebase I can try adapting it.

Reply:
Yeah, I'm pretty sure `KeepOutputValues: false` is what we want here. Any inputs that have dependencies will be tracked by the engine for the custom resource. They won't show up in the custom resource's outputs, but that shouldn't matter because any uses of the custom resource's outputs will have the custom resource as a dependency, so the transitive dependencies should all be tracked correctly by the engine.

Note if there are tests in this repo that are explicitly checking that `Module` outputs have dependencies from the inputs, those tests will have to be updated.

We should confirm the end-behavior we want works as expected, e.g. deletes happen in the right order.